Prometheus

mgoodness opened the pull request:
This is a fairly significant rework of the Prometheus chart. Goal is to provide decent defaults while allowing for thorough customization. Comments on my approach are _extremely_ welcome.

Changes/additions include:
- Separate Pods for Alertmanager & Server (including resource requests & limits)
- Separate ConfigMaps for Alertmanager & Server
  - All entries come from (and are overridable by) `values.yaml`
- Separate (optional) Ingresses for Alertmanager & Server
- Use [configmap-reload](https://github.com/jimmidyson/configmap-reload) to trigger config reloads
- Customizable annotations on Server Pod and both Ingresses
- Customizable args for Server container

TODO:
- [x] User-provided TLS Secrets
- [x] Documentation in `values.yaml` and `README`
- [x] `NOTES.txt`
- [x] Testing

Commit:
Prometheus rework

davidopp commented:
@fgrzadkowski @piosz

viglesiasce commented:
Awesome stuff @mgoodness! I'm going to do some thorough review tomorrow. If you get the NOTES.txt done in this PR we can migrate this to stable.

fgrzadkowski commented:
@fabxc FYI

Commit:
Optional Ingress TLS

mgoodness commented:
At this point I think the chart is functionally complete. It works on my cluster!

I'll begin documenting values, but would sure appreciate a review before I get too far along. Hit me up in Slack for any real-time discussion.

Commit:
Values.yaml documentation

lachie83 commented:
I will review this today.

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+
+## Configmap-reload Docker image
+##
+configmap_reload_image: jimmidyson/configmap-reload:v0.1
```
Probably worth calling this out as a source -- https://github.com/jimmidyson/configmap-reload

lachie83 commented on incubator/prometheus/templates/alertmanager-deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+          configMap:
+            name: {{ template "alertmanager.fullname" . }}
+        - name: storage-volume
+          emptyDir: {}
```
Should this be a PV? What's the impact of loosing this?

mgoodness replied:
Really good question. I was under the impression that AlertManager's storage was more of a cache, but [this post](https://groups.google.com/forum/#!topic/prometheus-developers/KQ5UbAbaYnU) would seem to indicate otherwise. I'll create a PV.

lachie83 replied:
I agree with your findings after reading that link.

lachie83 commented on incubator/prometheus/templates/alertmanager-ingress.yaml:
```diff
@@ -0,0 +1,31 @@
+    - host: {{ .hostName }}
+      http:
+        paths:
+          - path: /
```
Should this path be configurable? Would you ever want the prom server and the alert mgt behind the same ingress with a different path?

mgoodness replied:
Maybe. I recall trying that approach myself & having some difficulty. Can certainly give it another try, but IMHO shouldn't be a blocker.

lachie83 replied:
I'm happy to leave it for now. This is absolutely a non-blocker. Happy to leave it out of this PR unless someone has a compelling reason

lachie83 commented on incubator/prometheus/templates/alertmanager-deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+    release: "{{ .Release.Name }}"
+  name: {{ template "alertmanager.fullname" . }}
+spec:
+  replicas: 1
```
Should this be configurable? What's the impact of losing this Pod?

mgoodness replied:
[Alertmanager HA](https://github.com/prometheus/alertmanager) is under construction, and (if I understand correctly) may require a StatefulSet at that point. For now I think a PV will suffice.

lachie83 replied:
Agreed. Absolutely non-blocker. Lets address once Alertmanager has a compelling HA story.

lachie83 commented on incubator/prometheus/templates/alertmanager-svc.yaml:
```diff
@@ -0,0 +1,19 @@
+      protocol: TCP
+      targetPort: 9093
+  selector:
+    app: {{ template "fullname" . }}
```
Trying to get my head around the purpose of having both selectors? Deploying two prometheus + alert mgr deployments + svcs in the same namespace? Is that something I should test?

mgoodness replied:
I give server & alertmanager resources their own `component` label, but they share the `app` label. That allows for something like `kubectl get pod,svc,ing -l app=dining-wasp-prometheus` for easy retrieval of all `app` resources, while ensuring the services only select their `component`.

I hadn't considered the side-effect of this approach being able to run multiple deployments/services in the same namespace, but I can envision someone wanting that. Prometheus's HA strategy is basically "run two".

lachie83 replied:
Perfect. Thanks for the explanation. I might just adopt doing that myself.

lachie83 commented on incubator/prometheus/templates/alertmanager-svc.yaml:
```diff
@@ -0,0 +1,19 @@
+      name: {{ default "http" .Values.alertmanager.httpPortName | quote }}
+      protocol: TCP
+      targetPort: 9093
+  selector:
```
Is it worth having type LoadBalancer available as an configurable option in the case where you don't have ingress controllers and want to expose the service externally?

mgoodness replied:
Good call. Certainly doesn't hurt to provide the option.

lachie83 commented on incubator/prometheus/templates/alertmanager-deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+            - name: config-volume
+              mountPath: /etc/config
+            - name: storage-volume
+              mountPath: /data
```
Should this be `{{ default "/data" .Values.alertmanager.storage_path }}`?

mgoodness replied:
Nice catch! Also renamed the key for consistency.

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+
+## Alertmanager ConfigMap entries
+##
+alertmanager_files:
```
I would expect to find these under the alertmanager scope. Any reason they aren't there? The same goes for server_files under server scope respectively.

mgoodness replied:
I separated them because in a production setup the lists of files (as well as the files themselves) may grow quite long, and I think there's value in having a clear break between K8s resource config and internal component config.

lachie83 replied:
Makes sense. Appreciate the additional context

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
-  enabled: true
-  storageClass: generic
-  size: 8Gi
+alertmanager:
```
This is really nicely done. Love the formatting, documentation and configurable options. 👏

lachie83 commented on incubator/prometheus/templates/server-ingress.yaml:
```diff
@@ -0,0 +1,31 @@
+    - host: {{ .hostName }}
+      http:
+        paths:
+          - path: /
```
Should this be configurable? See alertmgr ingress for full question

mgoodness replied:
Same response as for alertmanager :)

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+    ## If true, Server will create a Persistent Volume Claim
+    ## If false, use emptyDir
+    ##
+    enabled: false
```
Why wouldn't you enable this by default?

mgoodness replied:
I guess because it assumes a PV provisioner will be available in the cluster. If the desire is to have truly production-ready defaults, then I completely agree this should be enabled by default.

lachie83 replied:
@viglesiasce Can you chime in on this? Should we go for prod ready defaults?

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+
+    ## If true, Server Ingress will be created
+    ##
+    enabled: false
```
I would put this directly underneath the ingress scope. It's easy to miss mid way

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+    ## If true, Server will create a Persistent Volume Claim
+    ## If false, use emptyDir
+    ##
+    enabled: false
```
I would put this directly under the persistentVolume scope

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+    #   - hostName: alertmanager.domain.com
+
+    ## Alertmanager Ingress TLS configuration
+    ## Secrets must be manually created in the namespace
```
Can you add a HOWTO on using TLS with this chart to the readme?

lachie83 commented on incubator/prometheus/templates/server-deployment.yaml:
```diff
@@ -0,0 +1,69 @@
+            - name: config-volume
+              mountPath: /etc/config
+            - name: storage-volume
+              mountPath: /data
```
Should this be `{{ default "/data" .Values.server.storage_local_path }}`

mgoodness replied:
Did it again. Thanks.

lachie83 commented on incubator/prometheus/templates/server-deployment.yaml:
```diff
@@ -0,0 +1,69 @@
+    release: "{{ .Release.Name }}"
+  name: {{ template "server.fullname" . }}
+spec:
+  replicas: 1
```
Should this be configurable?

mgoodness replied:
Each instance would require its own PV.

lachie83 replied:
Yeah and that's a job for Statefulsets. This is a non-blocker and shouldn't hold up this PR.

lachie83 reviewed (changes requested):
Great looking chart. 👍 Thanks for the contrib @mgoodness.

Here's my first pass at review.

jimmidyson commented on incubator/prometheus/values.yaml:
```diff
@@ -1,10 +1,200 @@
+## Server ConfigMap entries
+##
+server_files:
+  prometheus.yml: |
```
This is a really minimal config file, which is great but also makes it tricky for users to get started without learning Prometheus quite a bit. Most users I've come across are happy to start off with the example config in Prometheus repo (https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml) which includes a reasonable set of conventions (eg target configuration via annotations).

lachie83 replied:
+1

mgoodness replied:
This should have been on my TODO checklist. It does kind of demonstrate my point about keeping the K8s resource config separate from the server/alertmanager ConfigMap entries, though. Prometheus config gets verbose pretty quickly

jimmidyson commented:
cc @fabxc

Commit:
PR tweaks

* Consistent storage paths
* Configurable Alertmanager PV
* Configurable Service types
* Default Kubernetes scrape_configs

mgoodness commented:
Pretty sure I addressed all comments. Work continues on README and NOTES. Hopefully this weekend.

lachie83 commented:
I don't have too much Prometheus specific knowledge at this stage. @jimmidyson @fabxc could you kindly review the Prometheus runtime/config pieces?

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -118,6 +146,11 @@ server:
     #       - prometheus.domain.com
 
   persistentVolume:
```
@viglesiasce Any chance you can weigh in on this? Should we have data persistent enabled by default which would be more indicative of a prod-ready chart.

lachie83 commented on incubator/prometheus/values.yaml:
```diff
@@ -38,6 +38,28 @@ alertmanager:
+    ## If true, AlertManager will create a Persistent Volume Claim
+    ## If false, use emptyDir
+    ##
+    enabled: false
```
See other PV comment

lachie83 commented:
@mgoodness Updates looks great. Just an outstanding question on whether or not the PVs should be enabled be default

viglesiasce commented:
Since the replicas are currently hardcoded to 1 in each deployment we can enable the PVs by default. There will have to be some changes if we want to be able to support redundancy.

mgoodness commented:
I'll default-enable PVs.

`README` is coming along. Options are fully documented; just need to add instructions for enabling TLS on the Ingresses. Then I'll start on `NOTES.txt`.

viglesiasce commented:
I've been using the chart for the last few days and it rocks! Thanks @mgoodness!

Things that aren't working (for me):
- The default service target appears to be doing a label replace that ends up trying to scrape `http://blackbox:80/probe` which is not working. I found the line in the config which seems like its not quite right but also I dont know what the intent was.
- Its using port 80 for prometheus service which is a bit confusing since the default is 9090, took me a bit of poking to configure Grafana properly
- I havent yet had a successful config map reload despite multiple updates
- I couldn't get the alertmanager to fire but still looking into it

mgoodness commented:
I was probably a bit hasty in adding that default Prometheus config, and didn't check it as thoroughly as I should have. Apologies. The blackbox portion should come out since we don't install that exporter as part of the chart. I'll review the rest and make sure it's legit.

Will also double-check configmap-reload setup.

Prometheus 1.3.0 was released today and includes changes to K8s service discovery. I may just move the chart to that version.

jimmidyson commented:
@mgoodness I've just submitted an updated example config for 1.3.0 in https://github.com/prometheus/prometheus/pull/2148 if you want to upgrade - I think it's a good idea tbh (but remember to remove the example blackbox exporter part of the config if you're not using that).

jimmidyson commented:
@viglesiasce:

> I havent yet had a successful config map reload despite multiple updates

Are you sure the configmap had been updated in mounted volume? It's an async process that can take a while sometimes.

viglesiasce commented:
@jimmidyson you know I cant say that I ever waited "a while" before killing the pod myself. What is the cycle time? Mintues? Is there anything I can/should be looking for in the log?

mgoodness commented:
> > I havent yet had a successful config map reload despite multiple updates
> 
> Are you sure the configmap had been updated in mounted volume? It's an async process that can take a while sometimes.

Alternatively, if `configmap-reload` reports (through logs) that it's detecting the changes and triggering the webhook, it could be that the config is bad and Prometheus is ignoring it.

viglesiasce commented:
Just retried with removing the blackbox section and after about 2 minutes things reloaded as expected. Thanks for the pointers @jimmidyson @mgoodness. For reference here is what I saw in the logs:

``` console
2016/11/02 18:11:40 config map updated
2016/11/02 18:11:40 successfully triggered reload
```

viglesiasce commented:
I was also able to nail down why alerts werent firing. The alertmanager service is set to be hosted on 80 but 9093 is configured in the Prom deployment file. I say we set alertmanager (and prom) services to use their default ports.

jimmidyson commented:
> What is the cycle time?

I can't remember... it should be configurable too somewhere, cluster wide configuration (but I can't remember where!).

jimmidyson commented:
> I say we set alertmanager (and prom) services to use their default ports.

I personally prefer using the default proto ports (e.g. 80 for http services) but am not precious about it.

viglesiasce commented:
Yeah not a huge deal either way as long as they match up. I just got tripped up when setting up Grafana as I had expected to point it at the default Prom port.

viglesiasce reviewed (changes requested):
Needs alertmanager port set properly in server deployment

Commit:
Merge branch 'master' into prometheus

Commit:
Component labels, container naming, PVCs, AM port
  - Flexible component labels
  - Consistent container names
  - PVCs enabled by default
  - Commented blackbox exporter config
  - Fixed Alertmanager port

mgoodness commented:
Commented the blackbox exporter config, fixed the alertmanager port, enabled PVCs by default.

No 1.3.0 yet, nor HA Alertmanager (which was released recently). And I apparently lost all the changes I made to the `README`. So I have something to do on the way to & during KubeCon.

ptone commented:
One pattern I was just discussing with @viglesiasce was the use of an "extra args" iterator to allow people to inject arbitrary Prometheus flags (that don't require templatized values themselves)

just a proof of the idea for your consideration:
https://github.com/kubernetes/charts/compare/master...ptone:template-iteration?expand=1

mgoodness commented:
@ptone @viglesiasce I like that much more than feeding in raw YAML. Might try it in a few other places as well. Thanks!

Commit:
Use range when possible

Commit:
README & alertmanager PVC

Commit:
Prometheus 1.2.3, Alertmanager 0.5.0

Commit:
Added Ingress TLS info to README

Commit:
Added NOTES.txt

mgoodness commented:
Updated README and added NOTES. This chart is ready for stable, IMHO.

Commit:
NOTES tweaks

Commit:
Version 1.3.1

mgoodness commented:
@viglesiasce Edits enabled.

Commit:
Prometheus: Fix port-forwarding command in NOTES

Commit:
Prometheus: Set the annotations for storage

viglesiasce commented:
This LGTM. Would someone else mind giving it a test run?

lachie83 commented on incubator/prometheus/templates/NOTES.txt:
```diff
@@ -0,0 +1,74 @@
+  echo http://$SERVICE_IP:{{ .Values.server.httpPort }}
+{{- else if contains "ClusterIP"  .Values.server.serviceType }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "fullname" . }},component={{ .Values.server.name }}" -o jsonpath="{.items[0].metadata.name}")
+  kubectl port-forward $POD_NAME 9090
```
Should we add `--namespace {{ .Release.Namespace }}` to the port-forward command?

lachie83 commented on incubator/prometheus/templates/NOTES.txt:
```diff
@@ -0,0 +1,74 @@
+  echo http://$SERVICE_IP:{{ .Values.alertmanager.httpPort }}
+{{- else if contains "ClusterIP"  .Values.alertmanager.serviceType }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "fullname" . }},component={{ .Values.alertmanager.name }}" -o jsonpath="{.items[0].metadata.name}")
+  kubectl port-forward $POD_NAME {{ .Values.alertmanager.httpPort }}:{{ .Values.alertmanager.httpPort }}
```
Should we add `--namespace {{ .Release.Namespace }}` to the port-foward command?

Also the default .Values.alertmanager.httpPort is set to port 80 by default so we won't be able to bind to that locally without root. Should we use another port on the local side of the port forward?

mgoodness replied:
The `httpPort` is only used for the alertmanager service. The `containerPort` is 9093, so that's what we should use for port-forwarding.

lachie83 commented on incubator/prometheus/templates/NOTES.txt:
```diff
@@ -0,0 +1,74 @@
+  echo http://$NODE_IP:$NODE_PORT
+
+{{- else if contains "LoadBalancer" .Values.server.serviceType }}
+**** NOTE: It may take a few minutes for the LoadBalancer IP to be available.                             ****
```
Can you please remove the stars from NOTES.txt they end up getting rendered in a strange way.

lachie83 reviewed (commented):
Couple of ux cleanup items in NOTES.txt

Commit:
Added namespace to port-forward commands in NOTES

Commit:
Move Prometheus to stable

viglesiasce commented:
@k8s-bot ok to test

viglesiasce commented:
@k8s-bot ok to test

viglesiasce commented:
Looks like there is an issue with CI when moving things from incubator->stable. I tested this manually.

lachie83 commented:
LGTM
